### ISSUE TITLE
protoparse: options in descriptors are not serialized the same as with protoc

### ISSUE THREAD

**Issue by bufdev** [jhump/protoreflect#214]
```
    --- FAIL: TestEnvoyIncludeImports/envoy/config/filter/network/thrift_proxy/v2alpha1 (0.07s)
        require.go:1159:
            	Error Trace:	goprotoctest_test.go:639
            	            				goprotoctest_test.go:505
            	Error:      	Should be true
            	Test:       	TestEnvoyIncludeImports/envoy/config/filter/network/thrift_proxy/v2alpha1
            	Messages:   	diff detected:
            	            	protoc-goprotoc
            	            	File.slice[9].MessageType.slice[9].XXX_sizecache: 335 != 325
            	            	File.slice[9].XXX_sizecache: 2217 != 2207
            	            	XXX_sizecache: 38689 != 38679
```

A lot of these, all where the value of `XXX_sizecache` for `protoc` is 10 more than the value for `goprotoc`

**Comment by jhump** [comment]
Could you include more detail as to what this is your comparing? I'm not sure it's even valid to compare that field -- I think it's just a cache of the serialized size of the message and can change if the message is mutated after a serialization request. Was there _no other_ difference in the descriptors other than that field?

**Comment by pedgeio** [comment]
https://github.com/pedgeio/goprotoctest/blob/f52b200c1aa4b78eb68dad3b5ac0ba4dcc243063/internal/goprotoctest/goprotoctest_test.go#L122

This is directly comparing the `FileDescriptorSet` for every subdirectory in https://github.com/envoyproxy/envoy/tree/master/api. `protoc` is called to produce a `FileDescriptorSet` on a per-directory basis, then https://github.com/jhump/goprotoc/blob/master/app/goprotoc/goprotoc.go#L34 is called with the same arguments to produce a `FileDescriptorSet` as well, and the two are compared.

Note that:

- For https://github.com/googleapis/googleapis, there is no such difference in this field
- The message isn't changing after a serialization request

I think it is valid to compare them, even if irrelevant in most cases - it means `goprotoc` (and desc/protoparse) are producing different result. Also makes testing a lot harder to not have these equivalent in all cases.

**Comment by pedgeio** [comment]
This isn't just a Golang struct diff or a JSON diff FYI, this means there was a difference in the byte representation of the FileDescriptorSets https://github.com/pedgeio/goprotoctest/blob/f52b200c1aa4b78eb68dad3b5ac0ba4dcc243063/internal/goprotoctest/goprotoctest_test.go#L638

**Comment by pedgeio** [comment]
I did some more testing, it's very weird: `protoc` produces a `FileDescriptorSet` that is of size 10 more than the one produced by `goprotoc`, however both the JSON and String representation are the same.

**Comment by jhump** [comment]
Does `proto.Equal` say the messages are equal, but this one field differs? It seems very strange that the message might vary _only_ in this field, since it is not a field set by anything in goprotoc and is just set transparently by things in the runtime.

**Comment by pedgeio** [comment]
This is comparing the two `FileDescriptorSets` directly, this isn't relying on what the specific Golang implementation says is equal. It's not differing in any field, the actual serialized size of the two `FileDescriptorSets` is different, and they have slight byte differences, for example in one spot, instead of byte 169, there's byte 159 (neither are ASCII characters obviously).

Slight differences make testing tough and reduce the security of having `goprotoc` do at least what `protoc` does (and potentially more);.

**Comment by pedgeio** [comment]
It's actually more random than 10, this is difficult:

```
    --- FAIL: TestEnvoyIncludeImports/envoy/api/v2/core (0.05s)
        require.go:1159:
            	Error Trace:	goprotoctest_test.go:640
            	            				goprotoctest_test.go:506
            	Error:      	Should be true
            	Test:       	TestEnvoyIncludeImports/envoy/api/v2/core
            	Messages:   	diff detected:
            	            	protoc-goprotoc
            	            	File.slice[9].MessageType.slice[9].XXX_sizecache: 335 != 325
            	            	File.slice[9].XXX_sizecache: 2217 != 2207
            	            	File.slice[13].MessageType.slice[0].XXX_sizecache: 528 != 526
            	            	File.slice[13].XXX_sizecache: 1289 != 1287
            	            	File.slice[16].MessageType.slice[0].XXX_sizecache: 185 != 177
            	            	File.slice[16].XXX_sizecache: 379 != 371
            	            	XXX_sizecache: 33433 != 33413
```

I'm going through the files and I can't find a common pattern

**Comment by jhump** [comment]
Wait, so _every_ field in the two messages is the same _except_ that one field? While that is strange and worth getting to the bottom of, that certainly does not necessarily indicate a bug in this library.

Can you check whether `proto.Equal` in fact checks that they are the same? That method should be comparing them using protobuf semantics. Using something like `reflect.DeepEquals` isn't meant to be valid for protobufs.

In order to get to the bottom of this, can you try serializing both to bytes and see if the results differ? The size cache _should_ report the serialized size. So if it's working, the serialized outputs should differ in size. You could then feed the serialized outputs through `protoc --decode_raw` (or `goprotoc --decode_raw` 😛) and diff that to get a human-readable-ish diff of the binary outputs.

**Comment by pedgeio** [comment]
As I said and as I linked above, this is a byte comparison.

https://github.com/pedgeio/goprotoctest/blob/8a33d656fca5bc15e6cb78414444cea41a8bd970/internal/goprotoctest/goprotoctest_test.go#L638

**Comment by pedgeio** [comment]
I also just did a raw byte comparison at the source (ie the direct output of `--descriptor_set_out`, no serialize/deserialize in the middle), and the serialized `FileDescriptorSets` are in fact different sizes, in this case by 10 bytes.

```
protoc  36465
goprotoc  36455
```

**Comment by pedgeio** [comment]
The `sizecache` field being different is as you said, it's just indicative that the sizes are different for various messages (which I have confirmed they are)

**Comment by pedgeio** [comment]
OK I isolated it:

Options file:

```proto
// internal/options/options.go

syntax = "proto3";

import "google/protobuf/descriptor.proto";

package internal.options;

message Foo {
  string primitive = 1;
  string another_primitive = 2;
}

extend google.protobuf.FieldOptions {
  Foo field_foo_option = 50010;
}
```

Main file:

```proto
// foo.proto

syntax = "proto3";

package foo;

import "internal/options/options.proto";

message Bar {
  int64 eight = 8 [(internal.options.field_foo_option).primitive = "foo", (internal.options.field_foo_option).another_primitive = "bar"];
}
```

The problem happens when you have an option of a message type, and you specify it with `(option).one_field = "a", (option).two_field = "b"`, ie multiple fields within the message for the same option, but not specified in brackets. When this happens, desc/protoparse only sets the last value (so `(internal.options.field_foo_option).another_primitive = "bar"` is set), but drops the first value (and probably all but the first value).

This is the spot this happened in Envoy: https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/core/base.proto#L237

**Comment by pedgeio** [comment]
I think the issue is here: https://github.com/jhump/protoreflect/blob/master/desc/protoparse/options.go#L1277 - I think instead of always creating a new message, you need to see if one already exists. But I might be reading the code incorrectly, I only did a quick search

**Comment by jhump** [comment]
Hmmm. That is not quite the right part of the code -- this is:
https://github.com/jhump/protoreflect/blob/master/desc/protoparse/options.go#L1073

I also had stuff in testprotos that was intending to exercise this:
https://github.com/jhump/protoreflect/blob/master/internal/testprotos/desc_test_complex.proto#L87-L93

So I'll have to look a little more closely at what's going on.

**Comment by pedgeio** [comment]
I committed the test so you can reproduce it: https://github.com/pedgeio/goprotoctest/blob/74d991ee9dc6c4e1011d734aab1799079f196a54/internal/goprotoctest/goprotoctest_test.go#L59

The files are at:

https://github.com/pedgeio/goprotoctest/blob/74d991ee9dc6c4e1011d734aab1799079f196a54/internal/goprotoctest/testdata/internal/options/options.proto
https://github.com/pedgeio/goprotoctest/blob/74d991ee9dc6c4e1011d734aab1799079f196a54/internal/goprotoctest/testdata/field-options/foo.proto

**Comment by pedgeio** [comment]
The only differences immediately apparent to me there are:

- The option is declared in the same file in `desc_test_complex.proto` (but this is likely OK)\
- The option is a message option in `desc_test_complex.proto` (but this is likely OK)
- The option values being set in `desc_test_complex.proto` are all messages themselves, whereas this example is setting primitives on a message type - this doesn't seem like it would be it, but a quick test would be to make a FieldOption that has a message type, that message having two primitives, and setting it on a field in `desc_test_complex.proto`

**Comment by jhump** [comment]
I'm pretty sure I've figured out the issue and understand the differences. It can easily be reproduced with this sample file:

```proto
// test.proto
syntax = "proto3";

import "google/protobuf/descriptor.proto";

extend google.protobuf.FieldOptions {
  Foo foo = 20000;
}

message Foo {
  string bar = 1;
  string baz = 2;
  bool ok = 3;
}

message Msg {
  string id = 1 [(foo).bar = "frob", (foo).baz = "nitz", (foo).ok = true];
}
```

If I run `protoc test.proto -o test.protoset`, the resulting file is 224 bytes. But if I did the same thing using `desc/protoparse` package (like using `goprotoc`), the output file is just 216 bytes.

Interestingly, if I try to decode the resulting files, the output is _identical_!

```bash
protoc test.proto google/protobuf/descriptor.proto \
    --decode google.protobuf.FileDescriptorSet \
    < test.protoset
```

In particular, the resulting output for the options in question looks like so:
```
options {
  [foo] {
    bar: "frob"
    baz: "nitz"
    ok: true
  }
}
```

The true nature of the difference is only revealed by examining the actual encoded bytes:

```bash
protoc --decode_raw < test.protoset
```

Here, we see the difference between the two output descriptor sets. The one generated by `protoc` emits each option declaration as if it were a separate option (even though they are just separate fields all on the same option). Here's the relevant part of the decoded output:

```
8 {
  20000 {
    1: "frob"
  }
  20000 {
    2: "nitz"
  }
  20000 {
    3: 1
  }
}
```

However, the `desc/protoparse` package parses the options into a single message (since that option, at tag 20000, is not a repeated field) and then outputs it all in one:

```
8 {
  20000 {
    1: "frob"
    2: "nitz"
    3: 1
  }
}
```

So the file produced by `protoc` is larger because it repeats the tag+wire-type pre-amble for each option.

**Comment by pedgeio** [comment]
Re: our offline conversation, I went to test this out, and unfortunately I don't have the best news.

What I did: I produced binary FileDescriptorSets with both `goprotoc` and `protoc` for https://github.com/envoyproxy/envoy/tree/master/api/envoy/api/v2/core, outputting to files named `goprotoc.bin` and `protoc.bin`. Then, I put each through `--decode_raw`:

```
cat goprotoc.bin | protoc --decode_raw > goprotoc.txt
cat protoc.bin | protoc --decode_raw > protoc.txt
```

Then I diffed those files:

```
diff goprotoc.txt protoc.txt
```

This produced the following:

```diff
4022a4023,4027
>           17 {
>             2: 1
>           }
>         }
>         919191 {
4839c4844,4846
<             5: ""
---
>             5 {
>               1: 0
>             }
5529d5535
<             1: 1
5532a5539,5543
>         919191 {
>           21 {
>             1: 1
>           }
>         }
```

What is happening:

1. The `goprotoc` version does drop the first value of https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/core/base.proto#L238 - we thought we'd see a combined value for `919191` for the `goprotoc` version, but this isn't there.

```
// goprotoc.txt
    2 {
      1: "state"
      3: 6
      4: 1
      5: 14
      6: ".envoy.api.v2.core.SocketOption.SocketState"
      8 {
        919191 {
          16 {
            2: 1
          }
        }
      }
      10: "state"
    }
```

```
// protoc.txt
    2 {
      1: "state"
      3: 6
      4: 1
      5: 14
      6: ".envoy.api.v2.core.SocketOption.SocketState"
      8 {
        919191 {
          17 {
            2: 1
          }
        }
        919191 {
          16 {
            2: 1
          }
        }
      }
      10: "state"
    }
```

2. It drops values for https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/core/config_source.proto#L62

```
// goprotoc.txt
    2 {
      1: "request_timeout"
      3: 5
      4: 1
      5: 11
      6: ".google.protobuf.Duration"
      8 {
        65011: 1
        919191 {
          21 {
            5: ""
          }
        }
      }
      10: "requestTimeout"
    }
```

```
// protoc.txt
    2 {
      1: "request_timeout"
      3: 5
      4: 1
      5: 11
      6: ".google.protobuf.Duration"
      8 {
        65011: 1
        919191 {
          21 {
            5 {
              1: 0
            }
          }
        }
      }
      10: "requestTimeout"
    }
```

3. It drops values for https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/core/http_uri.proto#L47

```
// goprotoc.txt
    2 {
      1: "timeout"
      3: 3
      4: 1
      5: 11
      6: ".google.protobuf.Duration"
      8 {
        65011: 1
        919191 {
          21 {
            1: 1
            6: ""
          }
        }
      }
      10: "timeout"
    }
```

```
// protoc.txt
    2 {
      1: "timeout"
      3: 3
      4: 1
      5: 11
      6: ".google.protobuf.Duration"
      8 {
        65011: 1
        919191 {
          21 {
            6: ""
          }
        }
        919191 {
          21 {
            1: 1
          }
        }
      }
      10: "timeout"
    }
```

I've uploaded the full text files to https://gist.github.com/pedgeio/5dd975f24a41f918d5048d45805e3488 so you can see them.

It seems that there's something else going on - the bug still happens and it's not just combining as you'd expect, `goprotoc` is actually dropping the values.

**Comment by jhump** [comment]
The first of those three examples does appear to be an issue. But the others do not.

In the second example, it renders the duration as `5: ""` because it is an empty message. When decoding from raw bytes (no schema), it is not possible to distinguish an empty message from an empty string. It in fact _should_ be an empty message because the value is explicitly set to its zero value. And the file is "proto3" (this is `duration.proto`). The spec states that a value explicitly set to its zero value, in proto3, is indistinguishable from the field being absent. So this might be a bug in proto3.

In the third example, the values _are_ in fact combined in `goprotoc`:

```
// goprotoc
919191 {
  21 {
    1: 1
    6: ""
  }
}

// protoc
919191 {
  21 {
    6: ""
  }
}
919191 {
  21 {
    1: 1
  }
}
```

So I'll dig a little more into the first example and post an issue/question to the protobuf project about the second.

**Comment by pedgeio** [comment]
Yep I edited the issue, the third one is OK :-)

The second one I do not think is ok - if you have:

```proto
message Foo {
  int64 one = 1;
}

message Bar {
  Foo foo = 1; 
}
```

The `Bar` messages `{foo: {}}` and `{foo: nil}` ARE considered different. It's why if you need to check if a value is set or unset, you can use https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/wrappers.proto instead of primitives.

In this case, it does matter - https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/core/config_source.proto#L62 has a `google.protobuf.Duration`, which has signed integer values. This is making sure that the duration is non-negative.

**Comment by jhump** [comment]
That's not what's going on though. A more accurate example would be `Bar` messages `{foo: {}}` vs. `{foo: {bar: 0}}` which are _not_ considered different. In the case you describe, both are `Duration` messages with a zero value inside it.

**Comment by pedgeio** [comment]
Those two are not considered different yes, but case 2 above is - it has a different value, unless I'm missing something. It appears to be a null message, not an empty message.

**Comment by jhump** [comment]
> It appears to be a null message, not an empty message.

Why do you say that? The raw output of interest:

```
// goprotoc
919191 {
  21 {
     5: ""
  }
}

// protoc
919191 {
  21 {
    5 {
      1: 0
    }
  }
}
```

The message in question is tag 5. Goprotoc emits an empty message (`""`) where protoc emits an explicit zero-value for tag 1. Since the default value for tag 1 is zero, and the message is proto3, these are the same.

A `nil` would show up in the raw output as tag 5 simply missing from the goprotoc output.

**Comment by pedgeio** [comment]
Ah I was missing something :-)

This probably falls into the same category as case 3 then, where it would be nice to match `protoc` completely, but case 1 and #212 (and https://github.com/jhump/goprotoc/issues/17) take precedence.

**Comment by jhump** [comment]
FYI, of those three cases above -- where I said the first was a real issue but the others boiled down to the same symptom I had described earlier -- I did more digging. And... it is actually _also_ the same symptom. It just manifests weirdly.

This actually should probably result in a link error from both `protoc` _and_ `goprotoc`. What happened is that the two fields being set were together in a oneof (actually [fixed](https://github.com/envoyproxy/protoc-gen-validate/pull/213) shortly after you observed this issue)!

So `goprotoc` output was really the same symptom: it merged all of the options into a single message before writing out the descriptor. And since they were both in a oneof together, setting the second option _unset_ the first one. That's why it appeared to lose the first option.

I suspect that particularly case should actually be rejected since the option's are not valid together but are mutually exclusive. I'll file a bug with protoc about it, since it also incorrectly accepted the proto.
